### ISSUE TITLE
Refactor client factory to read remote cluster secrets from filesystem

### ISSUE THREAD

**Issue by nrfox** [kiali/kiali#5767]
Follow up from #5752. The client factory should be able to read each one of the remote secrets and construct a kube service account client from that. This functionality currently sits in the mesh service but this issue is to move that functionality into the client factory and expose the cluster info on the client interface so that the mesh service and other services can get the cluster info that way without needing to bother with loading in secrets or the kiali service account token.

**Comment by jmazzitelli** [comment]
WIP PRs:
* https://github.com/kiali/kiali/pull/5778
* https://github.com/kiali/kiali-operator/pull/614
* https://github.com/kiali/helm-charts/pull/191

**Comment by jmazzitelli** [comment]
The PRs are ready for review. This doesn't get us all the way there, but it does refactor the client factory to be multi-cluster aware. There is more work to be done, but the changes in here are substantial so we need this reviewed and tested and then put into master. This way we have a good checkpoint and we will avoid big git conflicts in the future (because this is changing alot of things).

**Comment by jmazzitelli** [comment]
After my own testing, I realize we cannot rely on the secrets generated by [`istioctl x create-remote-secret`](https://istio.io/latest/docs/reference/commands/istioctl/#istioctl-experimental-create-remote-secret) - those by default utilize the `istio-reader-service-account` which does not provide the permissions Kiali needs. Indeed, if you do tell kiali to use one of those secrets, the Kiali cache will fail to initialize when trying to load resources from the remote cluster, and Kiali will not start up. Even if we try to get that istioctl command to use some other SA, this will still break Kiali because those remote secrets that are used for istio to connect to the remote cluster will still be in the control plane namespace and Kiali will auto-detect them and try to use them and then break again.

We will need to provide our own tool to create Kiali remote secrets (I'm in the processes of writing a script to do just this).

**Comment by jmazzitelli** [comment]
Some notes that are relevant here.

In k8s 1.24+, secret-based tokens are no longer auto-created by default for new service accounts (ephemeral time-bound tokens is preferred in 1.21+ (see the TokenRequest API) if possible).

If a secret-based token is still desired, one can be created manually but will not be referenced from the service account's .secrets list. Create a secret with type `kubernetes.io/service-account-token` and the annotation `kubernetes.io/service-account.name` set to the SA name associated with the secret.

See https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-a-service-account-api-token

**Comment by israel-hdez** [comment]
> We will need to provide our own tool to create Kiali remote secrets (I'm in the processes of writing a script to do just this).

I was thinking to, perhaps, take advantage that we are already asking people to use helm. Maybe we can add a few small charts to generate the needed secrets, rather than giving some CLI tool. People tend to appreciate that they can review what is created in their clusters, rather than allowing a CLI tool to do everything on their behalf.

**Comment by jmazzitelli** [comment]
> I was thinking to, perhaps, take advantage that we are already asking people to use helm.

I'm kinda using that now with this script I'm putting together.
See the new hack/istio/kiali-create-remote-cluster-secret.sh in this PR - it is using our helm charts to build the necessary roles (and you can pass in `--view-only=true` for the restricted roles).

**Comment by jmazzitelli** [comment]
To be clear what needs to be done (whether by pure helm, or a script, or a combination of the two).

In the remote cluster:
* Create a role, rolebinding, SA, and a secret for the SA (in later k8s implementations, you do not get secrets anymore)

In the cluster where Kiali is to be installed
* Create a secret that contains a kubeconfig - that kubeconfig must have (1) the certificate authority for the remote cluster and (2) the token of the SA secret that was created in the remote cluster mentioned above.

So there are several things that need to be created in both clusters - the remote cluster and the kiali cluster.

**Comment by jmazzitelli** [comment]
> People tend to appreciate that they can review what is created in their clusters

This is a good option to have in any scripts we prepare. Helm provides that with --dry-run. I'll add something like that also to this hack script I'm putting together.

**Comment by jmazzitelli** [comment]
I have things working. Start a multicluster environ using the `install-everything.sh` script and it will create the remote cluster secrets for you. You should see this in the "east" Kiali:

![image](https://user-images.githubusercontent.com/2029470/217668106-9a79757d-8898-438e-a384-242b1efe0e7b.png)

And this in the logs shows it loads the secret:
```
DBG Data for remote cluster [west] has been loaded from secret
file [/kiali-remote-cluster-secrets/kiali-remote-cluster-secret-west/west]
```

**Comment by jmazzitelli** [comment]
How to test:

1. Make sure you checkout [the 3 PRs](https://github.com/kiali/kiali/issues/5767#issuecomment-1409429817) - server, operator, helm chart - in your typical Kiali dev environment set up.
2. Clean and build the helm charts:
  -- `cd <your helm chart repo> && make clean build-helm-charts`
3. Build the server container locally - this is the dev image that will be used in the install-everything.sh script later.
  -- `cd <your server repo> && make clean build build-ui container-build-kiali`
  -- Confirm you have a new `quay.io/kiali/kiali:dev` image that the above container-build-kiali target built
      ```
      $ podman images quay.io/kiali/kiali:dev
      REPOSITORY           TAG         IMAGE ID      CREATED        SIZE
      quay.io/kiali/kiali  dev         8d1e2c5932ad  3 seconds ago  108 MB
      ```
4.  Run the install-everything.sh script to start up 2 minikube clusters. This is all you should have to run. It will even run the new hack script to create the resources needed for the remote cluster secret that Kiali will use.
   -- This is all you should have to run. It will even run the new hack script to create the resources needed for the remote cluster secret that Kiali will use.
   -- The `-kudi true` option tells the script to push that dev image you created in step 3 into the east cluster and to install Kiali so it uses that dev container image.
      ```
      hack/istio/multicluster/install-everything.sh -c kubectl -kudi true
      ```

At this point, you can connect to Kiali at `https://192.168.59.252/kiali` where that IP is the IP of your minikube east cluster (you can get that via `minikube -p east ip`).

The Keycloak app can be connected at `https://keycloak-192-168-59-252.nip.io/` where that "192-168-59-252" is the east cluster's IP with the dots (.) replaced with dashes (-).

If you get odd Keycloak/OpenID errors, try with a private/incognito window or completely clear your browser's cache/cookies.

**Comment by josunect** [comment]
I had an issue before following these steps: 

`Error: UPGRADE FAILED: template: kiali-server/templates/deployment.yaml:130:33: executing "kiali-server/templates/deployment.yaml" at <include "kiali-server.remote-cluster-secrets" .>: error calling include: template: kiali-server/templates/_helpers.tpl:186:31: executing "kiali-server.remote-cluster-secrets" at <$secret.metadata.labels>: wrong type for value; expected map[string]interface {}; got interface {}`

That was fixed after Helm charts upgrade (Now using 3.11.1)

**Comment by jmazzitelli** [comment]
> `wrong type for value; expected map[string]interface {}; got interface {}`

Yes! This is the error I remember seeing with an old version of helm, and why I upgraded helm, too.
